File: tm_sync/training_load.py
"""Uitlegbare multi-sport belasting voor Garmin-activiteiten.

Garmin Training Load blijft de voorkeursbron. Ontbreekt die, dan gebruiken we
eerst hartslagzones en pas als laatste een expliciet gemarkeerde duurschatting.
Hardloopkilometers worden hier nooit met andere sporten vermengd.
"""

from __future__ import annotations

import json
import statistics
from collections import Counter
from datetime import date, timedelta
from pathlib import Path
from typing import Any

MODEL = json.loads(Path(__file__).with_name("training_load_model.json").read_text())
ZONE_WEIGHTS = tuple(float(value) for value in MODEL["zone_weights"])
MINIMUM_CHRONIC_LOAD = float(MODEL["minimum_chronic_load"])

SPORT_ALIASES = {
    "running": ("run", "jog", "trail", "treadmill"),
    "cycling": ("cycl", "bike", "biking", "bmx", "mountain_bik"),
    "swimming": ("swim",),
    "walking": ("walk",),
    "hiking": ("hik", "trek"),
    "strength": ("strength", "weight", "gym", "resistance", "functional"),
    "racquet": ("padel", "tennis", "squash", "badminton", "pickleball", "racquet"),
    "team_sport": ("football", "soccer", "basketball", "volleyball", "hockey", "rugby"),
    "rowing": ("row", "kayak", "canoe", "paddle"),
    "winter_sport": ("ski", "snowboard", "skate"),
    "yoga": ("yoga", "pilates", "mobility", "breathwork"),
}


def canonical_sport(sport: str | None, sub_sport: str | None = None, name: str | None = None) -> str:
    haystack = "_".join(str(value or "").lower().replace(" ", "_") for value in (sport, sub_sport, name))
    # Hiking moet voor walking worden gecontroleerd omdat Garmin soms beide
    # termen in hetzelfde subtype stopt.
    order = ("hiking", "running", "cycling", "swimming", "walking", "strength", "racquet", "team_sport", "rowing", "winter_sport", "yoga")
    for canonical in order:
        if any(alias in haystack for alias in SPORT_ALIASES[canonical]):
            return canonical
    return "other"


def _number(value: Any) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if number >= 0 else None


def estimate_activity_load(activity: dict[str, Any], zones: dict[int, float] | list[float] | tuple[float, ...] | None = None) -> dict[str, Any]:
    canonical = canonical_sport(activity.get("sport"), activity.get("sub_sport"), activity.get("name"))
    profile = MODEL["sports"][canonical]
    raw = activity.get("raw") or {}
    garmin = _number(raw.get("activityTrainingLoad"))

    zone_values: list[float] = [0.0] * 5
    if isinstance(zones, dict):
        for zone, seconds in zones.items():
            if 1 <= int(zone) <= 5:
                zone_values[int(zone) - 1] = _number(seconds) or 0.0
    elif zones:
        for index, seconds in enumerate(zones[:5]):
            zone_values[index] = _number(seconds) or 0.0
    zone_load = sum((seconds / 60.0) * ZONE_WEIGHTS[index] for index, seconds in enumerate(zone_values))

    duration_s = _number(activity.get("duration_s")) or 0.0
    if garmin and garmin > 0:
        load, source, confidence = garmin, "garmin", "high"
    elif zone_load > 0:
        load, source, confidence = zone_load, "heart_rate", "medium"
    elif duration_s > 0:
        load = duration_s / 60.0 * float(profile["load_per_minute"])
        source, confidence = "duration", "low"
    else:
        load, source, confidence = 0.0, "missing", "missing"

    return {
        "sport": str(activity.get("sport") or "other"),
        "canonical_sport": canonical,
        "load": round(load, 1),
        "aerobic_load": round(load * float(profile["aerobic_factor"]), 1),
        "mechanical_load": round(load * float(profile["mechanical_factor"]), 1),
        "source": source,
        "confidence": confidence,
        "estimated": source == "duration",
    }


def build_load_summary(activities: list[dict[str, Any]], zones_by_activity: dict[int, Any], end_day: date) -> dict[str, Any]:
    start_day = end_day - timedelta(days=27)
    current_start = end_day - timedelta(days=6)
    recent_start = end_day - timedelta(days=1)
    days = {
        (start_day + timedelta(days=index)).isoformat(): {
            "day": (start_day + timedelta(days=index)).isoformat(),
            "load": 0.0,
            "aerobic_load": 0.0,
            "mechanical_load": 0.0,
        }
        for index in range(28)
    }
    sports: dict[str, dict[str, Any]] = {}
    sources: Counter[str] = Counter()
    recent_cross = {"load": 0.0, "aerobic_load": 0.0, "mechanical_load": 0.0}

    for activity in activities:
        day = str(activity.get("start_time_local") or "")[:10]
        if day not in days:
            continue
        estimate = estimate_activity_load(activity, zones_by_activity.get(int(activity.get("id") or 0)))
        point = days[day]
        for key in ("load", "aerobic_load", "mechanical_load"):
            point[key] += estimate[key]
        sources[estimate["source"]] += 1

        if day >= current_start.isoformat():
            key = estimate["canonical_sport"]
            slot = sports.setdefault(key, {
                "sport": key,
                "load": 0.0,
                "aerobic_load": 0.0,
                "mechanical_load": 0.0,
                "duration_s": 0,
                "sessions": 0,
                "estimated_sessions": 0,
            })
            for field in ("load", "aerobic_load", "mechanical_load"):
                slot[field] += estimate[field]
            slot["duration_s"] += round(_number(activity.get("duration_s")) or 0)
            slot["sessions"] += 1
            slot["estimated_sessions"] += int(estimate["estimated"])

        if day >= recent_start.isoformat() and estimate["canonical_sport"] != "running":
            for field in recent_cross:
                recent_cross[field] += estimate[field]

    day_rows = list(days.values())
    for point in day_rows:
        for key in ("load", "aerobic_load", "mechanical_load"):
            point[key] = round(point[key], 1)
    current = sum(point["load"] for point in day_rows[-7:])
    previous = sum(point["load"] for point in day_rows[-14:-7])
    chronic = sum(point["load"] for point in day_rows) / 4.0
    ratio = current / chronic if chronic >= MINIMUM_CHRONIC_LOAD else None
    current_daily = [point["load"] for point in day_rows[-7:]]
    spread = statistics.pstdev(current_daily)
    monotony = statistics.fmean(current_daily) / spread if spread > 0 and any(current_daily) else None
    baseline_day = max(chronic / 7.0, 20.0)
    if recent_cross["load"] >= baseline_day * 2 or recent_cross["mechanical_load"] >= baseline_day:
        heavy_run_impact = "protect"
    elif recent_cross["load"] >= baseline_day or recent_cross["mechanical_load"] >= baseline_day * 0.5:
        heavy_run_impact = "watch"
    else:
        heavy_run_impact = "clear"

    measured = sources["garmin"] + sources["heart_rate"]
    estimated = sources["duration"]
    usable = measured + estimated
    if usable == 0:
        data_quality = "missing"
    elif estimated == 0:
        data_quality = "measured"
    elif measured > 0:
        data_quality = "mixed"
    else:
        data_quality = "estimated"

    return {
        "current_load": round(current, 1),
        "previous_load": round(previous, 1),
        "chronic_load": round(chronic, 1),
        "acwr": round(ratio, 2) if ratio is not None else None,
        "monotony": round(monotony, 2) if monotony is not None else None,
        "strain": round(current * monotony, 1) if monotony is not None else None,
        "days": day_rows,
        "sports": sorted(({
            **slot,
            "load": round(slot["load"], 1),
            "aerobic_load": round(slot["aerobic_load"], 1),
            "mechanical_load": round(slot["mechanical_load"], 1),
        } for slot in sports.values()), key=lambda item: item["load"], reverse=True),
        "recent_cross_load": {key: round(value, 1) for key, value in recent_cross.items()},
        "heavy_run_impact": heavy_run_impact,
        "data_quality": data_quality,
        "sources": dict(sources),
    }

File: web/lib/training-load.ts
import modelJson from "../../tm_sync/training_load_model.json";

type SportProfile = {
  load_per_minute: number;
  aerobic_factor: number;
  mechanical_factor: number;
};

const MODEL = modelJson as {
  version: number;
  minimum_chronic_load: number;
  zone_weights: number[];
  sports: Record<string, SportProfile>;
};

const SPORT_ALIASES: Record<string, string[]> = {
  running: ["run", "jog", "trail", "treadmill"],
  cycling: ["cycl", "bike", "biking", "bmx", "mountain_bik"],
  swimming: ["swim"],
  walking: ["walk"],
  hiking: ["hik", "trek"],
  strength: ["strength", "weight", "gym", "resistance", "functional"],
  racquet: ["padel", "tennis", "squash", "badminton", "pickleball", "racquet"],
  team_sport: ["football", "soccer", "basketball", "volleyball", "hockey", "rugby"],
  rowing: ["row", "kayak", "canoe", "paddle"],
  winter_sport: ["ski", "snowboard", "skate"],
  yoga: ["yoga", "pilates", "mobility", "breathwork"],
};

const SPORT_LABELS: Record<string, string> = {
  running: "Hardlopen", cycling: "Fietsen", swimming: "Zwemmen",
  walking: "Wandelen", hiking: "Hiken", strength: "Kracht",
  racquet: "Racketsport", team_sport: "Teamsport", rowing: "Roeien & peddelen",
  winter_sport: "Wintersport", yoga: "Yoga & mobiliteit", other: "Overige sport",
};

export function trainingLoadSportLabel(sport: string) {
  return SPORT_LABELS[sport] ?? sport.replaceAll("_", " ");
}

export type LoadSource = "garmin" | "heart_rate" | "duration" | "missing";
export type LoadDataQuality = "measured" | "mixed" | "estimated" | "missing";
export type HeavyRunImpact = "clear" | "watch" | "protect";

export type LoadActivityRow = {
  id: number;
  sport: string;
  sub_sport: string | null;
  name: string | null;
  start_time_local: string;
  duration_s: number | null;
  raw: { activityTrainingLoad?: number | string | null } | null;
};

export type TrainingLoadDay = {
  day: string;
  load: number;
  aerobicLoad: number;
  mechanicalLoad: number;
};

export type TrainingLoadSport = {
  sport: string;
  load: number;
  aerobicLoad: number;
  mechanicalLoad: number;
  duration_s: number;
  sessions: number;
  estimatedSessions: number;
};

export type TrainingLoadSource = {
  id: number;
  sport: string;
  sub_sport: string | null;
  name: string | null;
  start_time_local: string;
  duration_s: number;
  load: number | null;
  loadSource?: LoadSource;
};

export type TrainingLoadSummary = {
  currentLoad: number;
  previousLoad: number;
  chronicLoad: number;
  acwr: number | null;
  monotony: number | null;
  strain: number | null;
  deltaPct: number | null;
  currentDurationS: number;
  days: TrainingLoadDay[];
  sports: TrainingLoadSport[];
  lastActivity: TrainingLoadSource | null;
  recentCrossLoad: {
    load: number;
    aerobicLoad: number;
    mechanicalLoad: number;
  };
  heavyRunImpact: HeavyRunImpact;
  dataQuality: LoadDataQuality;
  sources: Record<LoadSource, number>;
};

export function canonicalSport(sport?: string | null, subSport?: string | null, name?: string | null) {
  const haystack = [sport, subSport, name]
    .map((value) => String(value ?? "").toLowerCase().replaceAll(" ", "_"))
    .join("_");
  const order = [
    "hiking", "running", "cycling", "swimming", "walking", "strength",
    "racquet", "team_sport", "rowing", "winter_sport", "yoga",
  ];
  return order.find((key) => SPORT_ALIASES[key].some((alias) => haystack.includes(alias))) ?? "other";
}

function rounded(value: number) {
  return Math.round(value * 10) / 10;
}

export function estimateActivityLoad(row: LoadActivityRow, zones: number[] = []) {
  const sport = canonicalSport(row.sport, row.sub_sport, row.name);
  const profile = MODEL.sports[sport] ?? MODEL.sports.other;
  const garminLoad = Number(row.raw?.activityTrainingLoad ?? 0);
  const zoneLoad = zones.reduce(
    (total, seconds, index) => total + (Number(seconds) / 60) * (MODEL.zone_weights[index] ?? index + 1),
    0,
  );
  const durationLoad = (Number(row.duration_s) || 0) / 60 * profile.load_per_minute;
  let load = 0;
  let source: LoadSource = "missing";
  if (Number.isFinite(garminLoad) && garminLoad > 0) {
    load = garminLoad;
    source = "garmin";
  } else if (zoneLoad > 0) {
    load = zoneLoad;
    source = "heart_rate";
  } else if (durationLoad > 0) {
    load = durationLoad;
    source = "duration";
  }
  return {
    sport,
    load: rounded(load),
    aerobicLoad: rounded(load * profile.aerobic_factor),
    mechanicalLoad: rounded(load * profile.mechanical_factor),
    source,
    estimated: source === "duration",
  };
}

function addDays(day: string, amount: number) {
  const value = new Date(`${day}T12:00:00Z`);
  value.setUTCDate(value.getUTCDate() + amount);
  return value.toISOString().slice(0, 10);
}

function pstdev(values: number[]) {
  if (!values.length) return 0;
  const mean = values.reduce((sum, value) => sum + value, 0) / values.length;
  return Math.sqrt(values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / values.length);
}

export function buildTrainingLoadSummary(
  rows: LoadActivityRow[],
  zones: Map<number, number[]>,
  endDay: string,
): TrainingLoadSummary {
  const fromDay = addDays(endDay, -27);
  const currentStart = addDays(endDay, -6);
  const recentStart = addDays(endDay, -1);
  const days = Array.from({ length: 28 }, (_, index): TrainingLoadDay => ({
    day: addDays(fromDay, index), load: 0, aerobicLoad: 0, mechanicalLoad: 0,
  }));
  const dayMap = new Map(days.map((day) => [day.day, day]));
  const sportMap = new Map<string, TrainingLoadSport>();
  const sources: Record<LoadSource, number> = { garmin: 0, heart_rate: 0, duration: 0, missing: 0 };
  const recentCrossLoad = { load: 0, aerobicLoad: 0, mechanicalLoad: 0 };
  let currentDurationS = 0;

  for (const row of rows) {
    const day = row.start_time_local.slice(0, 10);
    const estimate = estimateActivityLoad(row, zones.get(row.id));
    const point = dayMap.get(day);
    if (!point) continue;
    point.load += estimate.load;
    point.aerobicLoad += estimate.aerobicLoad;
    point.mechanicalLoad += estimate.mechanicalLoad;
    sources[estimate.source] += 1;

    if (day >= currentStart && day <= endDay) {
      const slot = sportMap.get(estimate.sport) ?? {
        sport: estimate.sport, load: 0, aerobicLoad: 0, mechanicalLoad: 0,
        duration_s: 0, sessions: 0, estimatedSessions: 0,
      };
      slot.load += estimate.load;
      slot.aerobicLoad += estimate.aerobicLoad;
      slot.mechanicalLoad += estimate.mechanicalLoad;
      slot.duration_s += Number(row.duration_s) || 0;
      slot.sessions += 1;
      slot.estimatedSessions += Number(estimate.estimated);
      currentDurationS += Number(row.duration_s) || 0;
      sportMap.set(estimate.sport, slot);
    }
    if (day >= recentStart && estimate.sport !== "running") {
      recentCrossLoad.load += estimate.load;
      recentCrossLoad.aerobicLoad += estimate.aerobicLoad;
      recentCrossLoad.mechanicalLoad += estimate.mechanicalLoad;
    }
  }

  days.forEach((day) => {
    day.load = rounded(day.load);
    day.aerobicLoad = rounded(day.aerobicLoad);
    day.mechanicalLoad = rounded(day.mechanicalLoad);
  });
  const currentLoad = days.slice(-7).reduce((sum, day) => sum + day.load, 0);
  const previousLoad = days.slice(-14, -7).reduce((sum, day) => sum + day.load, 0);
  const chronicLoad = days.reduce((sum, day) => sum + day.load, 0) / 4;
  const acwr = chronicLoad >= MODEL.minimum_chronic_load ? currentLoad / chronicLoad : null;
  const currentDaily = days.slice(-7).map((day) => day.load);
  const spread = pstdev(currentDaily);
  const monotony = spread > 0 && currentLoad > 0 ? currentLoad / 7 / spread : null;
  const baselineDay = Math.max(chronicLoad / 7, 20);
  const heavyRunImpact: HeavyRunImpact =
    recentCrossLoad.load >= baselineDay * 2 || recentCrossLoad.mechanicalLoad >= baselineDay
      ? "protect"
      : recentCrossLoad.load >= baselineDay || recentCrossLoad.mechanicalLoad >= baselineDay * 0.5
        ? "watch"
        : "clear";
  const measured = sources.garmin + sources.heart_rate;
  const usable = measured + sources.duration;
  const dataQuality: LoadDataQuality =
    usable === 0 ? "missing"
      : sources.duration === 0 ? "measured"
        : measured > 0 ? "mixed" : "estimated";
  const latest = rows.find((row) => row.start_time_local.slice(0, 10) >= currentStart) ?? null;
  const latestEstimate = latest ? estimateActivityLoad(latest, zones.get(latest.id)) : null;

  return {
    currentLoad: rounded(currentLoad),
    previousLoad: rounded(previousLoad),
    chronicLoad: rounded(chronicLoad),
    acwr: acwr == null ? null : Math.round(acwr * 100) / 100,
    monotony: monotony == null ? null : Math.round(monotony * 100) / 100,
    strain: monotony == null ? null : rounded(currentLoad * monotony),
    deltaPct: previousLoad > 0 ? Math.round(((currentLoad - previousLoad) / previousLoad) * 100) : null,
    currentDurationS: Math.round(currentDurationS),
    days,
    sports: [...sportMap.values()].map((sport) => ({
      ...sport,
      load: rounded(sport.load),
      aerobicLoad: rounded(sport.aerobicLoad),
      mechanicalLoad: rounded(sport.mechanicalLoad),
      duration_s: Math.round(sport.duration_s),
    })).sort((a, b) => b.load - a.load),
    lastActivity: latest && latestEstimate ? {
      id: latest.id, sport: latest.sport, sub_sport: latest.sub_sport, name: latest.name,
      start_time_local: latest.start_time_local, duration_s: Math.round(Number(latest.duration_s) || 0),
      load: latestEstimate.load, loadSource: latestEstimate.source,
    } : null,
    recentCrossLoad: {
      load: rounded(recentCrossLoad.load),
      aerobicLoad: rounded(recentCrossLoad.aerobicLoad),
      mechanicalLoad: rounded(recentCrossLoad.mechanicalLoad),
    },
    heavyRunImpact,
    dataQuality,
    sources,
  };
}
